Adopt std::span in more places

cdumez opened the pull request:
#### 5ec5a2c519e1e62d602d432ad243edf1641bd920
<pre>
Adopt std::span in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=272656">https://bugs.webkit.org/show_bug.cgi?id=272656</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/JSValueRef.cpp:
(JSValueMakeFromJSONString):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
(JSC::Interpreter::executeProgram):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseDecimalInt32):
(JSC::ISO8601::handleFraction):
(JSC::ISO8601::parseTimeSpec):
(JSC::ISO8601::parseDate):
* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::parseIndex):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::isInfinity):
(JSC::jsBinaryIntegerLiteral):
(JSC::jsOctalIntegerLiteral):
(JSC::jsHexIntegerLiteral):
(JSC::jsStrDecimalLiteral):
(JSC::toDouble):
(JSC::jsToNumber):
(JSC::parseFloat):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSONParse):
(JSC::JSONParseWithException):
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::LiteralParser):
(JSC::LiteralParser::Lexer::Lexer):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::transformCanLikelyUseFastPath):
(WebCore::parseSimpleTransformList):
(WebCore::parseSimpleTransform):
* Source/WebCore/dom/Document.cpp:
(WebCore::isValidNameNonASCII):
(WebCore::isValidNameASCII):
(WebCore::Document::isValidName):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::appendDescriptorAndReset):
(WebCore::parseImageCandidatesFromSrcsetAttribute):

Canonical link: <a href="https://commits.webkit.org/277508@main">https://commits.webkit.org/277508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/100c53fadfed8bff0d28a60f5dda75758017a06c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47791 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26987 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24470 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38899 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 27 flakes 75 failures; Uploaded test results; 14 flakes 59 failures; Running compile-webkit-without-change") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48373 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24662 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22139 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42497 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5840 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41079 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); Running jscore-test") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44160 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42918 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52368 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47284 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22828 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19190 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46205 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24100 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45246 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24890 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54782 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6764 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23821 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11250 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -390,12 +390,11 @@ JSValueRef JSValueMakeFromJSONString(JSContextRef ctx, JSStringRef string)
-    unsigned length = str.length();
-    if (!length || str.is8Bit()) {
-        LiteralParser<LChar> parser(globalObject, str.characters8(), length, StrictJSON);
+    if (str.isEmpty() || str.is8Bit()) {
```
Does `is8Bit` return true for empty strings?

cdumez replied:
Indeed, looking at the code, I believe the isEmpty() check is unnecessary.

darinadler commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -390,12 +390,11 @@ JSValueRef JSValueMakeFromJSONString(JSContextRef ctx, JSStringRef string)
-    if (!length || str.is8Bit()) {
-        LiteralParser<LChar> parser(globalObject, str.characters8(), length, StrictJSON);
+    if (str.isEmpty() || str.is8Bit()) {
+        LiteralParser<LChar> parser(globalObject, str.span8(), StrictJSON);
```
Can `LiteralParser` deduce the template argument?

cdumez replied:
Yes it can. I'll drop the template arguments.

darinadler commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -59,12 +59,12 @@ std::optional<TimeZoneID> parseTimeZoneName(StringView string)
-        result = (result * 10) + characters[index] - '0';
+    for (auto character : characters) {
+        ASSERT(isASCIIDigit(character));
+        result = (result * 10) + character - '0';
```
Seems like this just overflows. Is that really what we want? I guess callers always pass only short strings here. Surprised there isn’t a function elsewhere we can reuse.

darinadler commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -917,7 +917,7 @@ static std::optional<PlainDate> parseDate(StringParsingBuffer<CharacterType>& bu
                 return std::nullopt;
         }
-        year = parseDecimalInt32(buffer.position(), 6) * yearFactor;
+        year = parseDecimalInt32(std::span { buffer.position(), 6 }) * yearFactor;
```
Can we do this without `std::span`?

cdumez replied:
Doesn't build without it.

darinadler commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -928,7 +928,7 @@ static std::optional<PlainDate> parseDate(StringParsingBuffer<CharacterType>& bu
                 return std::nullopt;
         }
-        year = parseDecimalInt32(buffer.position(), 4);
+        year = parseDecimalInt32(std::span { buffer.position(), 4 });
```
Can we do this without `std::span`?

cdumez replied:
Doesn't build without it.

darinadler commented on Source/JavaScriptCore/runtime/Identifier.h:
```diff
@@ -50,17 +50,18 @@ ALWAYS_INLINE std::optional<uint32_t> parseIndex(const CharType* characters, uns
 
         // Get the next character, turning it into a digit.
-        uint32_t newValue = *(++characters) - '0';
+        uint32_t newValue = characters[0] - '0';
```
`characters.first()`

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -256,102 +256,102 @@ static bool isInfinity(const CharType* data, const CharType* end)
+        if (data.empty())
             break;
-        if (!isASCIIBinaryDigit(*data))
+        if (!isASCIIBinaryDigit(data[0]))
```
`data.front()`

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -256,102 +256,102 @@ static bool isInfinity(const CharType* data, const CharType* end)
-    data += 2;
-    const CharType* firstDigitPosition = data;
+    data = data.subspan(2);
+    auto* firstDigitPosition = data.data();
```
Use a `span` instead? It can remember both the position and how much of the data there is there for less overflow risk.

cdumez replied:
Ok

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -256,102 +256,102 @@ static bool isInfinity(const CharType* data, const CharType* end)
     }
     if (number >= mantissaOverflowLowerBound)
-        number = parseIntOverflow(firstDigitPosition, data - firstDigitPosition, 2);
+        number = parseIntOverflow(firstDigitPosition, data.data() - firstDigitPosition, 2);
```
Seems like `parseIntOverflow` should take a `span`.

cdumez replied:
Ok.

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -256,102 +256,102 @@ static bool isInfinity(const CharType* data, const CharType* end)
-        number = number * 2 + (*data - '0');
-        ++data;
-        if (data == end)
+        number = number * 2 + (data[0] - '0');
```
`data.front()`

darinadler reviewed (approved):
Comments apply to much similar code; I didn’t repeat them each time.

cdumez committed:
Adopt std::span in more places
https://bugs.webkit.org/show_bug.cgi?id=272656

Reviewed by Darin Adler.

* Source/JavaScriptCore/API/JSValueRef.cpp:
(JSValueMakeFromJSONString):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
(JSC::Interpreter::executeProgram):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseDecimalInt32):
(JSC::ISO8601::handleFraction):
(JSC::ISO8601::parseTimeSpec):
(JSC::ISO8601::parseDate):
* Source/JavaScriptCore/runtime/Identifier.h:
(JSC::parseIndex):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::isInfinity):
(JSC::jsBinaryIntegerLiteral):
(JSC::jsOctalIntegerLiteral):
(JSC::jsHexIntegerLiteral):
(JSC::jsStrDecimalLiteral):
(JSC::toDouble):
(JSC::jsToNumber):
(JSC::parseFloat):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSONParse):
(JSC::JSONParseWithException):
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::LiteralParser):
(JSC::LiteralParser::Lexer::Lexer):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::transformCanLikelyUseFastPath):
(WebCore::parseSimpleTransformList):
(WebCore::parseSimpleTransform):
* Source/WebCore/dom/Document.cpp:
(WebCore::isValidNameNonASCII):
(WebCore::isValidNameASCII):
(WebCore::Document::isValidName):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::appendDescriptorAndReset):
(WebCore::parseImageCandidatesFromSrcsetAttribute):

Canonical link: https://commits.webkit.org/277508@main
